src/nameservice/ClickNameService.sol

```diff
@@ -0,0 +1,194 @@
+import {Strings} from "openzeppelin-contracts/contracts/utils/Strings.sol";
+import {IClickNameService} from "./IClickNameService.sol";
+
+/**
+ * @title ClickNameService based on ClaveNameService which is authored by https://getclave.io
+ * @notice L2 name service contract that is built compatible to resolved as ENS subdomains by L2 resolver
+ * @dev Names can only be registered by authorized accounts
+ * @dev Addresses can only have one name at a time
+ * @dev Subdomains are stored as ERC-721 assets, cannot be transferred
+ * @dev If renewals are enabled, non-renewed names can be burnt after expiration timeline
+ */
+contract ClickNameService is IClickNameService, ERC721Burnable, AccessControl {
+    using Strings for uint256;
+
+    struct NameOwner {
+        address owner;
+        string name;
+    }
+
+    // Role to be authorized as default minter
+    bytes32 public constant REGISTERER_ROLE = keccak256("REGISTERER_ROLE");
+    // Default expiration duration
+    uint256 public expiryDuration = 365 days;
+
+    // token id to expires timestamp
+    mapping(uint256 => uint256) public expires;
+
+    event NameRegistered(string indexed name, address indexed owner, uint256 expires);
+
+    /// @notice Thrown when attempting to resolve a name that has expired
+    /// @param oldOwner The address of the previous owner of the name
+    /// @param expiredAt The timestamp when the name expired
+    error NameExpired(address oldOwner, uint256 expiredAt);
+
+    /// @notice Thrown when attempting to register a name that is not alphanumeric
+    error NameMustBeAlphanumeric();
+
+    /// @notice Thrown when attempting to register a name that is empty
+    error NameCannotBeEmpty();
+
+    /// @notice Thrown when attempting to change attributes of a name that is not found in the registry
+    error NameNotFound();
+
+    /// @notice Thrown when attempting to register a name that already exists
+    /// @param owner The address of the current owner of the name
+    /// @param expiresAt The timestamp when the name expires
+    error NamAlreadyExists(address owner, uint256 expiresAt);
+
+    /// @notice Thrown when attempting to remove a name that has not yet expired
+    /// @param expiresAt The timestamp when the name expires
+    error NameNotExpired(uint256 expiresAt);
+
+    /// @notice Thrown when an unauthorized address attempts to perform a permissioned action
+    error NotAuthorized();
+
+    constructor(address admin, address registrar) ERC721("ClickNameService", "CLK") {
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+        _grantRole(REGISTERER_ROLE, registrar);
+    }
+
+    /// @inheritdoc IClickNameService
+    function resolve(string memory name) external view returns (address) {
+        uint256 tokenId = uint256(keccak256(abi.encodePacked(name)));
+        address owner = ownerOf(tokenId);
+        if (expires[tokenId] <= block.timestamp) {
+            revert NameExpired(owner, expires[tokenId]);
+        }
```

Comment:
By default, names should be permanent unless we enable expiry via an admin function